File: server-go/internal/crypto/crypto.go
package crypto

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

func GenerateDeviceID() string {
	b := make([]byte, 16)
	rand.Read(b)
	return hex.EncodeToString(b)
}

func GeneratePairingKey(deviceID string) string {
	b := make([]byte, 8)
	rand.Read(b)
	hash := sha256.Sum256([]byte(deviceID + hex.EncodeToString(b)))
	return hex.EncodeToString(hash[:8])
}

func GenerateMessageID() string {
	b := make([]byte, 16)
	rand.Read(b)
	return hex.EncodeToString(b)
}

func GenerateFileCode() string {
	const charset = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
	b := make([]byte, 6)
	rand.Read(b)
	for i := range b {
		b[i] = charset[int(b[i])%len(charset)]
	}
	return string(b)
}

func HashPassword(password string) string {
	hash := sha256.Sum256([]byte(password))
	return hex.EncodeToString(hash[:])
}

func VerifyPasswordHash(password, hash string) bool {
	return HashPassword(password) == hash
}

func GenerateToken(deviceID string) string {
	b := make([]byte, 32)
	rand.Read(b)
	return fmt.Sprintf("%s:%x", deviceID, b)
}

File: server-rust/src/crypto.rs
use rand::Rng;
use sha2::{Digest, Sha256};

pub fn generate_device_id() -> String {
    let bytes: [u8; 16] = rand::thread_rng().gen();
    hex::encode(bytes)
}

pub fn generate_pairing_key(device_id: &str) -> String {
    let random_bytes: [u8; 8] = rand::thread_rng().gen();
    let mut hasher = Sha256::new();
    hasher.update(device_id.as_bytes());
    hasher.update(&random_bytes);
    let result = hasher.finalize();
    hex::encode(&result[..8])
}

pub fn generate_message_id() -> String {
    let bytes: [u8; 16] = rand::thread_rng().gen();
    hex::encode(bytes)
}

pub fn generate_file_code() -> String {
    const CHARSET: &[u8] = b"ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
    let mut rng = rand::thread_rng();
    let code: String = (0..6)
        .map(|_| {
            let idx = rng.gen_range(0..CHARSET.len());
            CHARSET[idx] as char
        })
        .collect();
    code
}

pub fn generate_token(device_id: &str) -> String {
    let bytes: [u8; 32] = rand::thread_rng().gen();
    format!("{}:{}", device_id, hex::encode(bytes))
}
